streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -336,15 +338,34 @@ private SinkNodeFactory(final String name,
             this.keySerializer = keySerializer;
             this.valSerializer = valSerializer;
             this.partitioner = partitioner;
+            this.topicExtractor = null;
+        }
+
+        private SinkNodeFactory(final String name,
+                                final String[] predecessors,
+                                final TopicNameExtractor<K, V> topicExtractor,
+                                final Serializer<K> keySerializer,
+                                final Serializer<V> valSerializer,
+                                final StreamPartitioner<? super K, ? super V> partitioner) {
+            super(name, predecessors.clone());
+            this.topicExtractor = topicExtractor;
+            this.keySerializer = keySerializer;
+            this.valSerializer = valSerializer;
+            this.partitioner = partitioner;
+            this.topic = null;
         }
 
         @Override
         public ProcessorNode build() {
-            if (internalTopicNames.contains(topic)) {
-                // prefix the internal topic name with the application id
-                return new SinkNode<>(name, decorateTopic(topic), keySerializer, valSerializer, partitioner);
+            if (topic != null) {
+                if (internalTopicNames.contains(topic)) {
```

Comment:
Maybe extract lines 361-363 to a method named something like `createSinkNodeWithTopic` returning a `SinkNode` then this method body would just have single `if-else` statement.  But not a big deal could stay as is.

Reply:
I think it is okay to keep as is since this function will only be called once so no harm to just inline it.